Rainbow accesses memory out-of-bounds

```
ur20980@uri-u20:~/src/PQClean/test$ uname -a && cc -v
Linux uri-u20 5.4.0-54-generic #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.3.0-17ubuntu1~20.04' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-HskZEa/gcc-9-9.3.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04) 
```


```
$ pytest -v -n=auto
.  .  .  .  .
cc -O3 -Wall -Wconversion -Wextra -Wpedantic -Wvla -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common -g3 -c -o rainbow_keypair_computation.o rainbow_keypair_computation.c
In file included from /usr/include/string.h:495,
                 from rainbow_keypair_computation.c:12:
In function ‘memset’,
    inlined from ‘calculate_Q_from_F_cyclic_ref’ at rainbow_keypair_computation.c:171:5:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:71:10: error: ‘__builtin_memset’ offset [153249, 188064] from the object at ‘tempQ’ is out of the bounds of referenced subobject ‘l1_F2’ with type ‘unsigned char[69632]’ at offset 83616 [-Werror=array-bounds]
   71 |   return __builtin___memset_chk (__dest, __ch, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [Makefile:14: rainbow_keypair_computation.o] Error 1
make[1]: Leaving directory '/home/ur20980/src/PQClean/testcases/test_valgrind__oWIB1B0U/crypto_bla/scheme/impl'
make: *** [Makefile:38: /home/ur20980/src/PQClean/testcases/test_valgrind__oWIB1B0U/crypto_bla/scheme/impl/librainbowIII-circumzenithal_clean.a] Error 2

_______________ test_valgrind[clean implementation of rainbowIII-compressed] _______________
[gw1] linux -- Python 3.8.5 /usr/bin/python3

implementation = <Implementation(rainbowIII-compressed, clean)>
impl_path = '/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/crypto_bla/scheme/impl'
test_dir = '/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX'
init = <function isolate_test_files.<locals>.initializer at 0x7f5f439e79d0>
destr = <function isolate_test_files.<locals>.destructor at 0x7f5f439e7a60>

    @pytest.mark.parametrize(
        'implementation,test_dir,impl_path, init, destr',
        [(impl, *helpers.isolate_test_files(impl.path(), 'test_valgrind_'))
         for impl in pqclean.Scheme.all_supported_implementations()],
        ids=[str(impl) for impl in pqclean.Scheme.all_supported_implementations()],
    )
    @helpers.slow_test
    @helpers.filtered_test
    def test_valgrind(implementation: pqclean.Implementation, impl_path, test_dir,
                      init, destr):
        if (platform.machine() not in ('i386', 'x86_64') or
                platform.system() != 'Linux'):
            raise unittest.SkipTest()
        init()
    
        dest_dir = os.path.join(test_dir, 'bin')
    
>       helpers.make(TYPE=implementation.scheme.type,
                     SCHEME=implementation.scheme.name,
                     SCHEME_DIR=os.path.abspath(impl_path),
                     IMPLEMENTATION=implementation.name,
                     DEST_DIR=dest_dir,
                     EXTRAFLAGS="-g3",
                     NTESTS=1,
                     working_dir=os.path.join(test_dir, 'test'))

test_valgrind.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
helpers.py:135: in make
    return run_subprocess(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

command = ['make', 'TYPE=sign', 'SCHEME=rainbowIII-compressed', 'SCHEME_DIR=/home/ur20980/src/PQClean/testcases/test_valgrind__j.../scheme/impl', 'IMPLEMENTATION=clean', 'DEST_DIR=/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/bin', ...]
working_dir = '/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/test', env = None
expected_returncode = 0, print_output = True
.  .  .  .  .
        if print_output:
            print(result.stdout.decode('utf-8'))
        if expected_returncode is not None:
>           assert result.returncode == expected_returncode, \
                "Got unexpected return code {}".format(result.returncode)
E           AssertionError: Got unexpected return code 2

helpers.py:110: AssertionError
----------------------------------- Captured stdout call -----------------------------------
/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/test > make TYPE=sign SCHEME=rainbowIII-compressed SCHEME_DIR=/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/crypto_bla/scheme/impl IMPLEMENTATION=clean DEST_DIR=/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/bin EXTRAFLAGS=-g3 NTESTS=1
cd /home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/crypto_bla/scheme/impl && make
make[1]: Entering directory '/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/crypto_bla/scheme/impl'
cc -O3 -Wall -Wconversion -Wextra -Wpedantic -Wvla -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common -g3 -c -o blas_comm.o blas_comm.c
cc -O3 -Wall -Wconversion -Wextra -Wpedantic -Wvla -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common -g3 -c -o parallel_matrix_op.o parallel_matrix_op.c
cc -O3 -Wall -Wconversion -Wextra -Wpedantic -Wvla -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common -g3 -c -o rainbow.o rainbow.c
cc -O3 -Wall -Wconversion -Wextra -Wpedantic -Wvla -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common -g3 -c -o rainbow_keypair.o rainbow_keypair.c
cc -O3 -Wall -Wconversion -Wextra -Wpedantic -Wvla -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common -g3 -c -o rainbow_keypair_computation.o rainbow_keypair_computation.c
In file included from /usr/include/string.h:495,
                 from rainbow_keypair_computation.c:12:
In function ‘memset’,
    inlined from ‘calculate_Q_from_F_cyclic_ref’ at rainbow_keypair_computation.c:171:5:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:71:10: error: ‘__builtin_memset’ offset [153249, 188064] from the object at ‘tempQ’ is out of the bounds of referenced subobject ‘l1_F2’ with type ‘unsigned char[69632]’ at offset 83616 [-Werror=array-bounds]
   71 |   return __builtin___memset_chk (__dest, __ch, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [Makefile:14: rainbow_keypair_computation.o] Error 1
make[1]: Leaving directory '/home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/crypto_bla/scheme/impl'
make: *** [Makefile:38: /home/ur20980/src/PQClean/testcases/test_valgrind__jVoMhXIX/crypto_bla/scheme/impl/librainbowIII-compressed_clean.a] Error 2

================================= short test summary info ==================================
FAILED test_compile_lib.py::test_compile_lib[clean implementation of rainbowV-compressed]
FAILED test_compile_lib.py::test_compile_lib[clean implementation of rainbowIII-circumzenithal]
FAILED test_compile_lib.py::test_compile_lib[clean implementation of rainbowV-circumzenithal]
FAILED test_compile_lib.py::test_compile_lib[clean implementation of rainbowIII-compressed]
FAILED test_dynamic_memory.py::test_dynamic_memory[clean implementation of rainbowV-compressed]
FAILED test_dynamic_memory.py::test_dynamic_memory[clean implementation of rainbowIII-circumzenithal]
FAILED test_dynamic_memory.py::test_dynamic_memory[clean implementation of rainbowV-circumzenithal]
FAILED test_dynamic_memory.py::test_dynamic_memory[clean implementation of rainbowIII-compressed]
FAILED test_functest.py::test_functest[clean implementation of rainbowV-compressed] - Ass...
FAILED test_functest.py::test_functest[clean implementation of rainbowIII-circumzenithal]
FAILED test_makefile_dependencies.py::test_makefile_dependencies[clean implementation of rainbowV-compressed]
FAILED test_makefile_dependencies.py::test_makefile_dependencies[clean implementation of rainbowIII-circumzenithal]
FAILED test_makefile_dependencies.py::test_makefile_dependencies[clean implementation of rainbowV-circumzenithal]
FAILED test_makefile_dependencies.py::test_makefile_dependencies[clean implementation of rainbowIII-compressed]
FAILED test_functest.py::test_functest[clean implementation of rainbowV-circumzenithal]
FAILED test_functest.py::test_functest[clean implementation of rainbowIII-compressed] - A...
FAILED test_nistkat.py::test_nistkat[clean implementation of rainbowV-compressed] - Asser...
FAILED test_nistkat.py::test_nistkat[clean implementation of rainbowIII-circumzenithal]
FAILED test_nistkat.py::test_nistkat[clean implementation of rainbowV-circumzenithal] - A...
FAILED test_nistkat.py::test_nistkat[clean implementation of rainbowIII-compressed] - Ass...
FAILED test_symbol_namespace.py::test_symbol_namespaces[clean implementation of rainbowV-compressed]
FAILED test_symbol_namespace.py::test_symbol_namespaces[clean implementation of rainbowIII-circumzenithal]
FAILED test_symbol_namespace.py::test_symbol_namespaces[clean implementation of rainbowV-circumzenithal]
FAILED test_symbol_namespace.py::test_symbol_namespaces[clean implementation of rainbowIII-compressed]
FAILED test_testvectors.py::test_testvectors[clean implementation of rainbowV-compressed]
FAILED test_testvectors.py::test_testvectors[clean implementation of rainbowIII-circumzenithal]
FAILED test_testvectors.py::test_testvectors[clean implementation of rainbowV-circumzenithal]
FAILED test_testvectors.py::test_testvectors[clean implementation of rainbowIII-compressed]
FAILED test_functest.py::test_functest_sanitizers[clean implementation of rainbowV-compressed]
FAILED test_functest.py::test_functest_sanitizers[clean implementation of rainbowIII-circumzenithal]
FAILED test_functest.py::test_functest_sanitizers[clean implementation of rainbowV-circumzenithal]
FAILED test_functest.py::test_functest_sanitizers[clean implementation of rainbowIII-compressed]
FAILED test_valgrind.py::test_valgrind[clean implementation of rainbowV-compressed] - Ass...
FAILED test_valgrind.py::test_valgrind[clean implementation of rainbowV-circumzenithal]
FAILED test_valgrind.py::test_valgrind[clean implementation of rainbowIII-circumzenithal]
FAILED test_valgrind.py::test_valgrind[clean implementation of rainbowIII-compressed] - A...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/usr/lib/python3.8/threading.py:306: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
================ 36 failed, 3987 passed, 337 skipped in 25725.55s (7:08:45) ================
```

On other platforms (MacOS and CentOS) tests pass...

--- a/crypto_sign/rainbowI-circumzenithal/clean/rainbow_keypair_computation.c
+++ b/crypto_sign/rainbowI-circumzenithal/clean/rainbow_keypair_computation.c
@@ -139,65 +139,65 @@ static void calculate_F_from_Q_ref(sk_t *Fs, const sk_t *Qs, sk_t *Ts) {
     batch_bmatTr_madd(Fs->l2_F6, Qs->l2_F2, _O1, Ts->t4, _V1, _V1_BYTE, _O2, _O2_BYTE);
 }
 
+#define _SIZE_BUFFER_F2 (_O2_BYTE * _V1 * _O2)
+#define _SIZE_BUFFER_F3 (_O2_BYTE * _V1 * _O2)
 static void calculate_Q_from_F_cyclic_ref(cpk_t *Qs, const sk_t *Fs, const sk_t *Ts) {
     // Layer 1: Computing Q5, Q3, Q6, Q9
 
     //  Q_pk.l1_F5s[i] = UT( T1tr* (F1 * T1 + F2) )
     const unsigned char *t2 = Ts->t4;
-    sk_t tempQ;
-    memcpy(tempQ.l1_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
-    batch_trimat_madd(tempQ.l1_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE); // F1*T1 + F2
-    memset(tempQ.l2_F1, 0, sizeof(tempQ.l2_F1));
-    memset(tempQ.l2_F2, 0, sizeof(tempQ.l2_F2));
-    batch_matTr_madd(tempQ.l2_F1, Ts->t1, _V1, _V1_BYTE, _O1, tempQ.l1_F2, _O1, _O1_BYTE); // T1tr*(F1*T1 + F2)
-    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
-    PQCLEAN_RAINBOWICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q5, tempQ.l2_F1, _O1, _O1_BYTE); // UT( ... )   // Q5
 
+    // assuming _O2 >= _O1
+
+    unsigned char buffer_F2[_SIZE_BUFFER_F2];
+    memcpy(buffer_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
+    batch_trimat_madd( buffer_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE);          // F1*T1 + F2
+
+    // assuming _O2 >= _O1
+    unsigned char buffer_F3[_SIZE_BUFFER_F3];
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, Ts->t1, _V1, _V1_BYTE, _O1, buffer_F2, _O1, _O1_BYTE);    // T1tr*(F1*T1 + F2) , release buffer_F2
+    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
+    PQCLEAN_RAINBOWICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q5, buffer_F3, _O1, _O1_BYTE);                          // UT( ... )   // Q5 , release buffer_F3
     /*
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
-          Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
-          Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
-          Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
-      */
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
+        Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
+        Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
+        Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
+    */
     memset(Qs->l1_Q3, 0, _O1_BYTE * _V1 * _O2);
     memset(Qs->l1_Q6, 0, _O1_BYTE * _O1 * _O2);
     memset(Qs->l1_Q9, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O2));
 
-    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F1*T2
-    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE); // F1_T2 + F2_T3
-
-    memset(tempQ.l1_F2, 0, _O1_BYTE * _V1 * _O2);                                    // should be F3. assuming: _O1 >= _O2
-    batch_matTr_madd(tempQ.l1_F2, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE); // T2tr * ( F1_T2 + F2_T3 )
-    PQCLEAN_RAINBOWICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q9, tempQ.l1_F2, _O2, _O1_BYTE);        // Q9
-
-    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE); // F1_F1T_T2 + F2_T3  // Q3
-
-    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);    // F2tr*T2
-    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE); // Q6
+    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);            // F1*T2
+    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE);      // F1_T2 + F
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE);      // T2tr *  ( F1_T2 + F2_T3 )
+    PQCLEAN_RAINBOWICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q9, buffer_F3, _O2, _O1_BYTE);                               // Q9 , release buffe
+    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);          // F1_F1T_T2 + F2_T3  /
+    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F2tr*T2
+    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE);    // Q6
     /*
-          Layer 2
-          Computing Q9:
-
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
-      */
-    sk_t tempQ2;
-    memcpy(tempQ2.l2_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);                              /// F3 actually.
-    batch_trimat_madd(tempQ2.l2_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);       // F1*T2 + F3
-    batch_mat_madd(tempQ2.l2_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F1_T2 + F2_T3 + F3
-
-    memset(tempQ.l2_F3, 0, _O2_BYTE * _V1 * _O2);
-    batch_matTr_madd(tempQ.l2_F3, t2, _V1, _V1_BYTE, _O2, tempQ2.l2_F3, _O2, _O2_BYTE); // T2tr * ( ..... )
-
-    memcpy(tempQ.l2_F6, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
-    batch_trimat_madd(tempQ.l2_F6, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F5*T3 + F6
-
-    batch_matTr_madd(tempQ.l2_F3, Ts->t3, _O1, _O1_BYTE, _O2, tempQ.l2_F6, _O2, _O2_BYTE); // T2tr*( ..... ) + T3tr*( ..... )
+        Layer 2
+        Computing
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
+    */
+    memcpy(buffer_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);           // F1*T2 + F3
+    batch_mat_madd(buffer_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);     // F1_T2 + F2_T3
+    memset(buffer_F2, 0, _O2_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F2, t2, _V1, _V1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);     // T2tr * ( ..... )  , release buffe
+    memcpy(buffer_F3, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);        // F5*T3 + F6
+    batch_matTr_madd(buffer_F2, Ts->t3, _O1, _O1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);   // T2tr*( ..... ) + T3tr*( ..... )
     memset(Qs->l2_Q9, 0, _O2_BYTE * N_TRIANGLE_TERMS(_O2));
-    PQCLEAN_RAINBOWICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l2_Q9, tempQ.l2_F3, _O2, _O2_BYTE); // Q9
+    PQCLEAN_RAINBOWICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l2_Q9, buffer_F2, _O2, _O2_BYTE);                                 // Q9
+    memset(buffer_F2, 0, _SIZE_BUFFER_F2);
+    memset(buffer_F3, 0, _SIZE_BUFFER_F3);
 }
 
 // Choosing implementations depends on the macros: _BLAS_SSE_ and _BLAS_AVX2_

--- a/crypto_sign/rainbowI-compressed/clean/rainbow_keypair_computation.c
+++ b/crypto_sign/rainbowI-compressed/clean/rainbow_keypair_computation.c
@@ -139,65 +139,65 @@ static void calculate_F_from_Q_ref(sk_t *Fs, const sk_t *Qs, sk_t *Ts) {
     batch_bmatTr_madd(Fs->l2_F6, Qs->l2_F2, _O1, Ts->t4, _V1, _V1_BYTE, _O2, _O2_BYTE);
 }
 
+#define _SIZE_BUFFER_F2 (_O2_BYTE * _V1 * _O2)
+#define _SIZE_BUFFER_F3 (_O2_BYTE * _V1 * _O2)
 static void calculate_Q_from_F_cyclic_ref(cpk_t *Qs, const sk_t *Fs, const sk_t *Ts) {
     // Layer 1: Computing Q5, Q3, Q6, Q9
 
     //  Q_pk.l1_F5s[i] = UT( T1tr* (F1 * T1 + F2) )
     const unsigned char *t2 = Ts->t4;
-    sk_t tempQ;
-    memcpy(tempQ.l1_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
-    batch_trimat_madd(tempQ.l1_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE); // F1*T1 + F2
-    memset(tempQ.l2_F1, 0, sizeof(tempQ.l2_F1));
-    memset(tempQ.l2_F2, 0, sizeof(tempQ.l2_F2));
-    batch_matTr_madd(tempQ.l2_F1, Ts->t1, _V1, _V1_BYTE, _O1, tempQ.l1_F2, _O1, _O1_BYTE); // T1tr*(F1*T1 + F2)
-    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
-    PQCLEAN_RAINBOWICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q5, tempQ.l2_F1, _O1, _O1_BYTE); // UT( ... )   // Q5
 
+    // assuming _O2 >= _O1
+
+    unsigned char buffer_F2[_SIZE_BUFFER_F2];
+    memcpy(buffer_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
+    batch_trimat_madd( buffer_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE);          // F1*T1 + F2
+
+    // assuming _O2 >= _O1
+    unsigned char buffer_F3[_SIZE_BUFFER_F3];
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, Ts->t1, _V1, _V1_BYTE, _O1, buffer_F2, _O1, _O1_BYTE);    // T1tr*(F1*T1 + F2) , release buffer_F2
+    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
+    PQCLEAN_RAINBOWICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q5, buffer_F3, _O1, _O1_BYTE);                          // UT( ... )   // Q5 , release buffer_F3
     /*
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
-          Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
-          Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
-          Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
-      */
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
+        Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
+        Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
+        Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
+    */
     memset(Qs->l1_Q3, 0, _O1_BYTE * _V1 * _O2);
     memset(Qs->l1_Q6, 0, _O1_BYTE * _O1 * _O2);
     memset(Qs->l1_Q9, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O2));
 
-    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F1*T2
-    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE); // F1_T2 + F2_T3
-
-    memset(tempQ.l1_F2, 0, _O1_BYTE * _V1 * _O2);                                    // should be F3. assuming: _O1 >= _O2
-    batch_matTr_madd(tempQ.l1_F2, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE); // T2tr * ( F1_T2 + F2_T3 )
-    PQCLEAN_RAINBOWICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q9, tempQ.l1_F2, _O2, _O1_BYTE);        // Q9
-
-    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE); // F1_F1T_T2 + F2_T3  // Q3
-
-    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);    // F2tr*T2
-    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE); // Q6
+    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);            // F1*T2
+    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE);      // F1_T2 + F
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE);      // T2tr *  ( F1_T2 + F2_T3 )
+    PQCLEAN_RAINBOWICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q9, buffer_F3, _O2, _O1_BYTE);                               // Q9 , release buffe
+    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);          // F1_F1T_T2 + F2_T3  /
+    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F2tr*T2
+    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE);    // Q6
     /*
-          Layer 2
-          Computing Q9:
-
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
-      */
-    sk_t tempQ2;
-    memcpy(tempQ2.l2_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);                              /// F3 actually.
-    batch_trimat_madd(tempQ2.l2_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);       // F1*T2 + F3
-    batch_mat_madd(tempQ2.l2_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F1_T2 + F2_T3 + F3
-
-    memset(tempQ.l2_F3, 0, _O2_BYTE * _V1 * _O2);
-    batch_matTr_madd(tempQ.l2_F3, t2, _V1, _V1_BYTE, _O2, tempQ2.l2_F3, _O2, _O2_BYTE); // T2tr * ( ..... )
-
-    memcpy(tempQ.l2_F6, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
-    batch_trimat_madd(tempQ.l2_F6, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F5*T3 + F6
-
-    batch_matTr_madd(tempQ.l2_F3, Ts->t3, _O1, _O1_BYTE, _O2, tempQ.l2_F6, _O2, _O2_BYTE); // T2tr*( ..... ) + T3tr*( ..... )
+        Layer 2
+        Computing
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
+    */
+    memcpy(buffer_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);           // F1*T2 + F3
+    batch_mat_madd(buffer_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);     // F1_T2 + F2_T3
+    memset(buffer_F2, 0, _O2_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F2, t2, _V1, _V1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);     // T2tr * ( ..... )  , release buffe
+    memcpy(buffer_F3, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);        // F5*T3 + F6
+    batch_matTr_madd(buffer_F2, Ts->t3, _O1, _O1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);   // T2tr*( ..... ) + T3tr*( ..... )
     memset(Qs->l2_Q9, 0, _O2_BYTE * N_TRIANGLE_TERMS(_O2));
-    PQCLEAN_RAINBOWICOMPRESSED_CLEAN_UpperTrianglize(Qs->l2_Q9, tempQ.l2_F3, _O2, _O2_BYTE); // Q9
+    PQCLEAN_RAINBOWICOMPRESSED_CLEAN_UpperTrianglize(Qs->l2_Q9, buffer_F2, _O2, _O2_BYTE);                                 // Q9
+    memset(buffer_F2, 0, _SIZE_BUFFER_F2);
+    memset(buffer_F3, 0, _SIZE_BUFFER_F3);
 }
 
 // Choosing implementations depends on the macros: _BLAS_SSE_ and _BLAS_AVX2_

--- a/crypto_sign/rainbowIII-circumzenithal/clean/rainbow_keypair_computation.c
+++ b/crypto_sign/rainbowIII-circumzenithal/clean/rainbow_keypair_computation.c
@@ -139,65 +139,65 @@ static void calculate_F_from_Q_ref(sk_t *Fs, const sk_t *Qs, sk_t *Ts) {
     batch_bmatTr_madd(Fs->l2_F6, Qs->l2_F2, _O1, Ts->t4, _V1, _V1_BYTE, _O2, _O2_BYTE);
 }
 
+#define _SIZE_BUFFER_F2 (_O2_BYTE * _V1 * _O2)
+#define _SIZE_BUFFER_F3 (_O2_BYTE * _V1 * _O2)
 static void calculate_Q_from_F_cyclic_ref(cpk_t *Qs, const sk_t *Fs, const sk_t *Ts) {
     // Layer 1: Computing Q5, Q3, Q6, Q9
 
     //  Q_pk.l1_F5s[i] = UT( T1tr* (F1 * T1 + F2) )
     const unsigned char *t2 = Ts->t4;
-    sk_t tempQ;
-    memcpy(tempQ.l1_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
-    batch_trimat_madd(tempQ.l1_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE); // F1*T1 + F2
-    memset(tempQ.l2_F1, 0, sizeof(tempQ.l2_F1));
-    memset(tempQ.l2_F2, 0, sizeof(tempQ.l2_F2));
-    batch_matTr_madd(tempQ.l2_F1, Ts->t1, _V1, _V1_BYTE, _O1, tempQ.l1_F2, _O1, _O1_BYTE); // T1tr*(F1*T1 + F2)
-    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
-    PQCLEAN_RAINBOWIIICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q5, tempQ.l2_F1, _O1, _O1_BYTE); // UT( ... )   // Q5
 
+    // assuming _O2 >= _O1
+
+    unsigned char buffer_F2[_SIZE_BUFFER_F2];
+    memcpy(buffer_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
+    batch_trimat_madd( buffer_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE);          // F1*T1 + F2
+
+    // assuming _O2 >= _O1
+    unsigned char buffer_F3[_SIZE_BUFFER_F3];
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, Ts->t1, _V1, _V1_BYTE, _O1, buffer_F2, _O1, _O1_BYTE);    // T1tr*(F1*T1 + F2) , release buffer_F2
+    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
+    PQCLEAN_RAINBOWIIICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q5, buffer_F3, _O1, _O1_BYTE);                          // UT( ... )   // Q5 , release buffer_F3
     /*
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
-          Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
-          Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
-          Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
-      */
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
+        Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
+        Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
+        Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
+    */
     memset(Qs->l1_Q3, 0, _O1_BYTE * _V1 * _O2);
     memset(Qs->l1_Q6, 0, _O1_BYTE * _O1 * _O2);
     memset(Qs->l1_Q9, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O2));
 
-    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F1*T2
-    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE); // F1_T2 + F2_T3
-
-    memset(tempQ.l1_F2, 0, _O1_BYTE * _V1 * _O2);                                    // should be F3. assuming: _O1 >= _O2
-    batch_matTr_madd(tempQ.l1_F2, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE); // T2tr * ( F1_T2 + F2_T3 )
-    PQCLEAN_RAINBOWIIICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q9, tempQ.l1_F2, _O2, _O1_BYTE);        // Q9
-
-    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE); // F1_F1T_T2 + F2_T3  // Q3
-
-    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);    // F2tr*T2
-    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE); // Q6
+    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);            // F1*T2
+    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE);      // F1_T2 + F
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE);      // T2tr *  ( F1_T2 + F2_T3 )
+    PQCLEAN_RAINBOWIIICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q9, buffer_F3, _O2, _O1_BYTE);                               // Q9 , release buffe
+    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);          // F1_F1T_T2 + F2_T3  /
+    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F2tr*T2
+    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE);    // Q6
     /*
-          Layer 2
-          Computing Q9:
-
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
-      */
-    sk_t tempQ2;
-    memcpy(tempQ2.l2_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);                              /// F3 actually.
-    batch_trimat_madd(tempQ2.l2_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);       // F1*T2 + F3
-    batch_mat_madd(tempQ2.l2_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F1_T2 + F2_T3 + F3
-
-    memset(tempQ.l2_F3, 0, _O2_BYTE * _V1 * _O2);
-    batch_matTr_madd(tempQ.l2_F3, t2, _V1, _V1_BYTE, _O2, tempQ2.l2_F3, _O2, _O2_BYTE); // T2tr * ( ..... )
-
-    memcpy(tempQ.l2_F6, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
-    batch_trimat_madd(tempQ.l2_F6, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F5*T3 + F6
-
-    batch_matTr_madd(tempQ.l2_F3, Ts->t3, _O1, _O1_BYTE, _O2, tempQ.l2_F6, _O2, _O2_BYTE); // T2tr*( ..... ) + T3tr*( ..... )
+        Layer 2
+        Computing
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
+    */
+    memcpy(buffer_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);           // F1*T2 + F3
+    batch_mat_madd(buffer_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);     // F1_T2 + F2_T3
+    memset(buffer_F2, 0, _O2_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F2, t2, _V1, _V1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);     // T2tr * ( ..... )  , release buffe
+    memcpy(buffer_F3, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);        // F5*T3 + F6
+    batch_matTr_madd(buffer_F2, Ts->t3, _O1, _O1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);   // T2tr*( ..... ) + T3tr*( ..... )
     memset(Qs->l2_Q9, 0, _O2_BYTE * N_TRIANGLE_TERMS(_O2));
-    PQCLEAN_RAINBOWIIICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l2_Q9, tempQ.l2_F3, _O2, _O2_BYTE); // Q9
+    PQCLEAN_RAINBOWIIICIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l2_Q9, buffer_F2, _O2, _O2_BYTE);                                 // Q9
+    memset(buffer_F2, 0, _SIZE_BUFFER_F2);
+    memset(buffer_F3, 0, _SIZE_BUFFER_F3);
 }
 
 // Choosing implementations depends on the macros: _BLAS_SSE_ and _BLAS_AVX2_

--- a/crypto_sign/rainbowIII-compressed/clean/rainbow_keypair_computation.c
+++ b/crypto_sign/rainbowIII-compressed/clean/rainbow_keypair_computation.c
@@ -139,65 +139,65 @@ static void calculate_F_from_Q_ref(sk_t *Fs, const sk_t *Qs, sk_t *Ts) {
     batch_bmatTr_madd(Fs->l2_F6, Qs->l2_F2, _O1, Ts->t4, _V1, _V1_BYTE, _O2, _O2_BYTE);
 }
 
+#define _SIZE_BUFFER_F2 (_O2_BYTE * _V1 * _O2)
+#define _SIZE_BUFFER_F3 (_O2_BYTE * _V1 * _O2)
 static void calculate_Q_from_F_cyclic_ref(cpk_t *Qs, const sk_t *Fs, const sk_t *Ts) {
     // Layer 1: Computing Q5, Q3, Q6, Q9
 
     //  Q_pk.l1_F5s[i] = UT( T1tr* (F1 * T1 + F2) )
     const unsigned char *t2 = Ts->t4;
-    sk_t tempQ;
-    memcpy(tempQ.l1_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
-    batch_trimat_madd(tempQ.l1_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE); // F1*T1 + F2
-    memset(tempQ.l2_F1, 0, sizeof(tempQ.l2_F1));
-    memset(tempQ.l2_F2, 0, sizeof(tempQ.l2_F2));
-    batch_matTr_madd(tempQ.l2_F1, Ts->t1, _V1, _V1_BYTE, _O1, tempQ.l1_F2, _O1, _O1_BYTE); // T1tr*(F1*T1 + F2)
-    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
-    PQCLEAN_RAINBOWIIICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q5, tempQ.l2_F1, _O1, _O1_BYTE); // UT( ... )   // Q5
 
+    // assuming _O2 >= _O1
+
+    unsigned char buffer_F2[_SIZE_BUFFER_F2];
+    memcpy(buffer_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
+    batch_trimat_madd( buffer_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE);          // F1*T1 + F2
+
+    // assuming _O2 >= _O1
+    unsigned char buffer_F3[_SIZE_BUFFER_F3];
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, Ts->t1, _V1, _V1_BYTE, _O1, buffer_F2, _O1, _O1_BYTE);    // T1tr*(F1*T1 + F2) , release buffer_F2
+    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
+    PQCLEAN_RAINBOWIIICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q5, buffer_F3, _O1, _O1_BYTE);                          // UT( ... )   // Q5 , release buffer_F3
     /*
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
-          Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
-          Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
-          Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
-      */
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
+        Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
+        Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
+        Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
+    */
     memset(Qs->l1_Q3, 0, _O1_BYTE * _V1 * _O2);
     memset(Qs->l1_Q6, 0, _O1_BYTE * _O1 * _O2);
     memset(Qs->l1_Q9, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O2));
 
-    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F1*T2
-    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE); // F1_T2 + F2_T3
-
-    memset(tempQ.l1_F2, 0, _O1_BYTE * _V1 * _O2);                                    // should be F3. assuming: _O1 >= _O2
-    batch_matTr_madd(tempQ.l1_F2, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE); // T2tr * ( F1_T2 + F2_T3 )
-    PQCLEAN_RAINBOWIIICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q9, tempQ.l1_F2, _O2, _O1_BYTE);        // Q9
-
-    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE); // F1_F1T_T2 + F2_T3  // Q3
-
-    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);    // F2tr*T2
-    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE); // Q6
+    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);            // F1*T2
+    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE);      // F1_T2 + F
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE);      // T2tr *  ( F1_T2 + F2_T3 )
+    PQCLEAN_RAINBOWIIICOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q9, buffer_F3, _O2, _O1_BYTE);                               // Q9 , release buffe
+    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);          // F1_F1T_T2 + F2_T3  /
+    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F2tr*T2
+    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE);    // Q6
     /*
-          Layer 2
-          Computing Q9:
-
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
-      */
-    sk_t tempQ2;
-    memcpy(tempQ2.l2_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);                              /// F3 actually.
-    batch_trimat_madd(tempQ2.l2_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);       // F1*T2 + F3
-    batch_mat_madd(tempQ2.l2_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F1_T2 + F2_T3 + F3
-
-    memset(tempQ.l2_F3, 0, _O2_BYTE * _V1 * _O2);
-    batch_matTr_madd(tempQ.l2_F3, t2, _V1, _V1_BYTE, _O2, tempQ2.l2_F3, _O2, _O2_BYTE); // T2tr * ( ..... )
-
-    memcpy(tempQ.l2_F6, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
-    batch_trimat_madd(tempQ.l2_F6, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F5*T3 + F6
-
-    batch_matTr_madd(tempQ.l2_F3, Ts->t3, _O1, _O1_BYTE, _O2, tempQ.l2_F6, _O2, _O2_BYTE); // T2tr*( ..... ) + T3tr*( ..... )
+        Layer 2
+        Computing
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
+    */
+    memcpy(buffer_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);           // F1*T2 + F3
+    batch_mat_madd(buffer_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);     // F1_T2 + F2_T3
+    memset(buffer_F2, 0, _O2_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F2, t2, _V1, _V1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);     // T2tr * ( ..... )  , release buffe
+    memcpy(buffer_F3, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);        // F5*T3 + F6
+    batch_matTr_madd(buffer_F2, Ts->t3, _O1, _O1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);   // T2tr*( ..... ) + T3tr*( ..... )
     memset(Qs->l2_Q9, 0, _O2_BYTE * N_TRIANGLE_TERMS(_O2));
-    PQCLEAN_RAINBOWIIICOMPRESSED_CLEAN_UpperTrianglize(Qs->l2_Q9, tempQ.l2_F3, _O2, _O2_BYTE); // Q9
+    PQCLEAN_RAINBOWIIICOMPRESSED_CLEAN_UpperTrianglize(Qs->l2_Q9, buffer_F2, _O2, _O2_BYTE);                                 // Q9
+    memset(buffer_F2, 0, _SIZE_BUFFER_F2);
+    memset(buffer_F3, 0, _SIZE_BUFFER_F3);
 }
 
 // Choosing implementations depends on the macros: _BLAS_SSE_ and _BLAS_AVX2_

--- a/crypto_sign/rainbowV-circumzenithal/clean/rainbow_keypair_computation.c
+++ b/crypto_sign/rainbowV-circumzenithal/clean/rainbow_keypair_computation.c
@@ -139,65 +139,65 @@ static void calculate_F_from_Q_ref(sk_t *Fs, const sk_t *Qs, sk_t *Ts) {
     batch_bmatTr_madd(Fs->l2_F6, Qs->l2_F2, _O1, Ts->t4, _V1, _V1_BYTE, _O2, _O2_BYTE);
 }
 
+#define _SIZE_BUFFER_F2 (_O2_BYTE * _V1 * _O2)
+#define _SIZE_BUFFER_F3 (_O2_BYTE * _V1 * _O2)
 static void calculate_Q_from_F_cyclic_ref(cpk_t *Qs, const sk_t *Fs, const sk_t *Ts) {
     // Layer 1: Computing Q5, Q3, Q6, Q9
 
     //  Q_pk.l1_F5s[i] = UT( T1tr* (F1 * T1 + F2) )
     const unsigned char *t2 = Ts->t4;
-    sk_t tempQ;
-    memcpy(tempQ.l1_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
-    batch_trimat_madd(tempQ.l1_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE); // F1*T1 + F2
-    memset(tempQ.l2_F1, 0, sizeof(tempQ.l2_F1));
-    memset(tempQ.l2_F2, 0, sizeof(tempQ.l2_F2));
-    batch_matTr_madd(tempQ.l2_F1, Ts->t1, _V1, _V1_BYTE, _O1, tempQ.l1_F2, _O1, _O1_BYTE); // T1tr*(F1*T1 + F2)
-    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
-    PQCLEAN_RAINBOWVCIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q5, tempQ.l2_F1, _O1, _O1_BYTE); // UT( ... )   // Q5
 
+    // assuming _O2 >= _O1
+
+    unsigned char buffer_F2[_SIZE_BUFFER_F2];
+    memcpy(buffer_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
+    batch_trimat_madd( buffer_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE);          // F1*T1 + F2
+
+    // assuming _O2 >= _O1
+    unsigned char buffer_F3[_SIZE_BUFFER_F3];
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, Ts->t1, _V1, _V1_BYTE, _O1, buffer_F2, _O1, _O1_BYTE);    // T1tr*(F1*T1 + F2) , release buffer_F2
+    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
+    PQCLEAN_RAINBOWVCIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q5, buffer_F3, _O1, _O1_BYTE);                          // UT( ... )   // Q5 , release buffer_F3
     /*
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
-          Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
-          Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
-          Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
-      */
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
+        Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
+        Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
+        Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
+    */
     memset(Qs->l1_Q3, 0, _O1_BYTE * _V1 * _O2);
     memset(Qs->l1_Q6, 0, _O1_BYTE * _O1 * _O2);
     memset(Qs->l1_Q9, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O2));
 
-    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F1*T2
-    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE); // F1_T2 + F2_T3
-
-    memset(tempQ.l1_F2, 0, _O1_BYTE * _V1 * _O2);                                    // should be F3. assuming: _O1 >= _O2
-    batch_matTr_madd(tempQ.l1_F2, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE); // T2tr * ( F1_T2 + F2_T3 )
-    PQCLEAN_RAINBOWVCIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q9, tempQ.l1_F2, _O2, _O1_BYTE);        // Q9
-
-    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE); // F1_F1T_T2 + F2_T3  // Q3
-
-    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);    // F2tr*T2
-    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE); // Q6
+    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);            // F1*T2
+    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE);      // F1_T2 + F
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE);      // T2tr *  ( F1_T2 + F2_T3 )
+    PQCLEAN_RAINBOWVCIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l1_Q9, buffer_F3, _O2, _O1_BYTE);                               // Q9 , release buffe
+    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);          // F1_F1T_T2 + F2_T3  /
+    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F2tr*T2
+    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE);    // Q6
     /*
-          Layer 2
-          Computing Q9:
-
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
-      */
-    sk_t tempQ2;
-    memcpy(tempQ2.l2_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);                              /// F3 actually.
-    batch_trimat_madd(tempQ2.l2_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);       // F1*T2 + F3
-    batch_mat_madd(tempQ2.l2_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F1_T2 + F2_T3 + F3
-
-    memset(tempQ.l2_F3, 0, _O2_BYTE * _V1 * _O2);
-    batch_matTr_madd(tempQ.l2_F3, t2, _V1, _V1_BYTE, _O2, tempQ2.l2_F3, _O2, _O2_BYTE); // T2tr * ( ..... )
-
-    memcpy(tempQ.l2_F6, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
-    batch_trimat_madd(tempQ.l2_F6, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F5*T3 + F6
-
-    batch_matTr_madd(tempQ.l2_F3, Ts->t3, _O1, _O1_BYTE, _O2, tempQ.l2_F6, _O2, _O2_BYTE); // T2tr*( ..... ) + T3tr*( ..... )
+        Layer 2
+        Computing
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
+    */
+    memcpy(buffer_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);           // F1*T2 + F3
+    batch_mat_madd(buffer_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);     // F1_T2 + F2_T3
+    memset(buffer_F2, 0, _O2_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F2, t2, _V1, _V1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);     // T2tr * ( ..... )  , release buffe
+    memcpy(buffer_F3, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);        // F5*T3 + F6
+    batch_matTr_madd(buffer_F2, Ts->t3, _O1, _O1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);   // T2tr*( ..... ) + T3tr*( ..... )
     memset(Qs->l2_Q9, 0, _O2_BYTE * N_TRIANGLE_TERMS(_O2));
-    PQCLEAN_RAINBOWVCIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l2_Q9, tempQ.l2_F3, _O2, _O2_BYTE); // Q9
+    PQCLEAN_RAINBOWVCIRCUMZENITHAL_CLEAN_UpperTrianglize(Qs->l2_Q9, buffer_F2, _O2, _O2_BYTE);                                 // Q9
+    memset(buffer_F2, 0, _SIZE_BUFFER_F2);
+    memset(buffer_F3, 0, _SIZE_BUFFER_F3);
 }
 
 // Choosing implementations depends on the macros: _BLAS_SSE_ and _BLAS_AVX2_

--- a/crypto_sign/rainbowV-compressed/clean/rainbow_keypair_computation.c
+++ b/crypto_sign/rainbowV-compressed/clean/rainbow_keypair_computation.c
@@ -139,65 +139,65 @@ static void calculate_F_from_Q_ref(sk_t *Fs, const sk_t *Qs, sk_t *Ts) {
     batch_bmatTr_madd(Fs->l2_F6, Qs->l2_F2, _O1, Ts->t4, _V1, _V1_BYTE, _O2, _O2_BYTE);
 }
 
+#define _SIZE_BUFFER_F2 (_O2_BYTE * _V1 * _O2)
+#define _SIZE_BUFFER_F3 (_O2_BYTE * _V1 * _O2)
 static void calculate_Q_from_F_cyclic_ref(cpk_t *Qs, const sk_t *Fs, const sk_t *Ts) {
     // Layer 1: Computing Q5, Q3, Q6, Q9
 
     //  Q_pk.l1_F5s[i] = UT( T1tr* (F1 * T1 + F2) )
     const unsigned char *t2 = Ts->t4;
-    sk_t tempQ;
-    memcpy(tempQ.l1_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
-    batch_trimat_madd(tempQ.l1_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE); // F1*T1 + F2
-    memset(tempQ.l2_F1, 0, sizeof(tempQ.l2_F1));
-    memset(tempQ.l2_F2, 0, sizeof(tempQ.l2_F2));
-    batch_matTr_madd(tempQ.l2_F1, Ts->t1, _V1, _V1_BYTE, _O1, tempQ.l1_F2, _O1, _O1_BYTE); // T1tr*(F1*T1 + F2)
-    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
-    PQCLEAN_RAINBOWVCOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q5, tempQ.l2_F1, _O1, _O1_BYTE); // UT( ... )   // Q5
 
+    // assuming _O2 >= _O1
+
+    unsigned char buffer_F2[_SIZE_BUFFER_F2];
+    memcpy(buffer_F2, Fs->l1_F2, _O1_BYTE * _V1 * _O1);
+    batch_trimat_madd( buffer_F2, Fs->l1_F1, Ts->t1, _V1, _V1_BYTE, _O1, _O1_BYTE);          // F1*T1 + F2
+
+    // assuming _O2 >= _O1
+    unsigned char buffer_F3[_SIZE_BUFFER_F3];
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, Ts->t1, _V1, _V1_BYTE, _O1, buffer_F2, _O1, _O1_BYTE);    // T1tr*(F1*T1 + F2) , release buffer_F2
+    memset(Qs->l1_Q5, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O1));
+    PQCLEAN_RAINBOWVCOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q5, buffer_F3, _O1, _O1_BYTE);                          // UT( ... )   // Q5 , release buffer_F3
     /*
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
-          Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
-          Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
-          Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
-      */
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        F1_F1T_T2 + F2_T3 = F1_T2 + F2_T3 + F1tr * t2
+        Q_pk.l1_F3s[i] =         F1_F1T_T2 + F2_T3
+        Q_pk.l1_F6s[i] = T1tr* ( F1_F1T_T2 + F2_T3 ) + F2tr * t2
+        Q_pk.l1_F9s[i] = UT( T2tr* ( F1_T2 + F2_T3 ) )
+    */
     memset(Qs->l1_Q3, 0, _O1_BYTE * _V1 * _O2);
     memset(Qs->l1_Q6, 0, _O1_BYTE * _O1 * _O2);
     memset(Qs->l1_Q9, 0, _O1_BYTE * N_TRIANGLE_TERMS(_O2));
 
-    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F1*T2
-    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE); // F1_T2 + F2_T3
-
-    memset(tempQ.l1_F2, 0, _O1_BYTE * _V1 * _O2);                                    // should be F3. assuming: _O1 >= _O2
-    batch_matTr_madd(tempQ.l1_F2, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE); // T2tr * ( F1_T2 + F2_T3 )
-    PQCLEAN_RAINBOWVCOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q9, tempQ.l1_F2, _O2, _O1_BYTE);        // Q9
-
-    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE); // F1_F1T_T2 + F2_T3  // Q3
-
-    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);    // F2tr*T2
-    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE); // Q6
+    batch_trimat_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);            // F1*T2
+    batch_mat_madd(Qs->l1_Q3, Fs->l1_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O1_BYTE);      // F1_T2 + F
+    memset(buffer_F3, 0, _O1_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F3, t2, _V1, _V1_BYTE, _O2, Qs->l1_Q3, _O2, _O1_BYTE);      // T2tr *  ( F1_T2 + F2_T3 )
+    PQCLEAN_RAINBOWVCOMPRESSED_CLEAN_UpperTrianglize(Qs->l1_Q9, buffer_F3, _O2, _O1_BYTE);                               // Q9 , release buffe
+    batch_trimatTr_madd(Qs->l1_Q3, Fs->l1_F1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);          // F1_F1T_T2 + F2_T3  /
+    batch_bmatTr_madd(Qs->l1_Q6, Fs->l1_F2, _O1, t2, _V1, _V1_BYTE, _O2, _O1_BYTE);       // F2tr*T2
+    batch_matTr_madd(Qs->l1_Q6, Ts->t1, _V1, _V1_BYTE, _O1, Qs->l1_Q3, _O2, _O1_BYTE);    // Q6
     /*
-          Layer 2
-          Computing Q9:
-
-          F1_T2     = F1 * t2
-          F2_T3     = F2 * t3
-          Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
-      */
-    sk_t tempQ2;
-    memcpy(tempQ2.l2_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);                              /// F3 actually.
-    batch_trimat_madd(tempQ2.l2_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);       // F1*T2 + F3
-    batch_mat_madd(tempQ2.l2_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F1_T2 + F2_T3 + F3
-
-    memset(tempQ.l2_F3, 0, _O2_BYTE * _V1 * _O2);
-    batch_matTr_madd(tempQ.l2_F3, t2, _V1, _V1_BYTE, _O2, tempQ2.l2_F3, _O2, _O2_BYTE); // T2tr * ( ..... )
-
-    memcpy(tempQ.l2_F6, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
-    batch_trimat_madd(tempQ.l2_F6, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE); // F5*T3 + F6
-
-    batch_matTr_madd(tempQ.l2_F3, Ts->t3, _O1, _O1_BYTE, _O2, tempQ.l2_F6, _O2, _O2_BYTE); // T2tr*( ..... ) + T3tr*( ..... )
+        Layer 2
+        Computing
+        F1_T2     = F1 * t2
+        F2_T3     = F2 * t3
+        Q9 = UT( T2tr*( F1*T2 + F2*T3 + F3 )  +  T3tr*( F5*T3 + F6 ) )
+    */
+    memcpy(buffer_F3, Fs->l2_F3, _O2_BYTE * _V1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F1, t2, _V1, _V1_BYTE, _O2, _O2_BYTE);           // F1*T2 + F3
+    batch_mat_madd(buffer_F3, Fs->l2_F2, _V1, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);     // F1_T2 + F2_T3
+    memset(buffer_F2, 0, _O2_BYTE * _V1 * _O2);
+    batch_matTr_madd(buffer_F2, t2, _V1, _V1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);     // T2tr * ( ..... )  , release buffe
+    memcpy(buffer_F3, Fs->l2_F6, _O2_BYTE * _O1 * _O2);
+    batch_trimat_madd(buffer_F3, Fs->l2_F5, Ts->t3, _O1, _O1_BYTE, _O2, _O2_BYTE);        // F5*T3 + F6
+    batch_matTr_madd(buffer_F2, Ts->t3, _O1, _O1_BYTE, _O2, buffer_F3, _O2, _O2_BYTE);   // T2tr*( ..... ) + T3tr*( ..... )
     memset(Qs->l2_Q9, 0, _O2_BYTE * N_TRIANGLE_TERMS(_O2));
-    PQCLEAN_RAINBOWVCOMPRESSED_CLEAN_UpperTrianglize(Qs->l2_Q9, tempQ.l2_F3, _O2, _O2_BYTE); // Q9
+    PQCLEAN_RAINBOWVCOMPRESSED_CLEAN_UpperTrianglize(Qs->l2_Q9, buffer_F2, _O2, _O2_BYTE);                                 // Q9
+    memset(buffer_F2, 0, _SIZE_BUFFER_F2);
+    memset(buffer_F3, 0, _SIZE_BUFFER_F3);
 }
 
 // Choosing implementations depends on the macros: _BLAS_SSE_ and _BLAS_AVX2_

